packages/alchemy/src/react/hooks/useSmartAccountClient.ts

```diff
@@ -0,0 +1,97 @@
+  type PluginManagerActions,
+} from "@alchemy/aa-accounts";
+import type { Chain, Transport } from "viem";
+import { createAlchemySmartAccountClientFromRpcClient } from "../../client/internal/smartAccountClientFromRpc.js";
+import type {
+  AlchemySmartAccountClient,
+  AlchemySmartAccountClientConfig,
+  BaseAlchemyActions,
+} from "../../client/smartAccountClient";
+import type { SupportedAccounts } from "../../config";
+import type { AlchemySigner } from "../../signer";
+import { useBundlerClient } from "./useBundlerClient.js";
+
+export type UseSmartAccountClientProps<
+  TTransport extends Transport = Transport,
+  TChain extends Chain | undefined = Chain | undefined,
+  TAccount extends SupportedAccounts = SupportedAccounts
+> = Omit<
+  AlchemySmartAccountClientConfig<TTransport, TChain, TAccount>,
+  "rpcUrl" | "chain" | "apiKey" | "jwt" | "account"
+> & { account: TAccount };
+
+export type ClientActions<
+  TChain extends Chain | undefined = Chain | undefined,
+  TAccount extends SupportedAccounts = SupportedAccounts
+> = TAccount extends LightAccount
+  ? BaseAlchemyActions<TChain, LightAccount<AlchemySigner>> &
+      LightAccountClientActions<AlchemySigner>
+  : TAccount extends MultiOwnerModularAccount
+  ? BaseAlchemyActions<TChain, MultiOwnerModularAccount<AlchemySigner>> &
+      MultiOwnerPluginActions<MultiOwnerModularAccount<AlchemySigner>> &
+      PluginManagerActions<MultiOwnerModularAccount<AlchemySigner>> &
+      AccountLoupeActions<MultiOwnerModularAccount<AlchemySigner>>
+  : never;
+
+export type UseSmartAccountClientResult<
+  TTransport extends Transport = Transport,
+  TChain extends Chain | undefined = Chain | undefined,
+  TAccount extends SupportedAccounts = SupportedAccounts
+> = AlchemySmartAccountClient<
+  TTransport,
+  TChain,
+  TAccount,
+  ClientActions<TChain, TAccount>
+>;
+
+export function useSmartAccountClient<
+  TTransport extends Transport = Transport,
+  TChain extends Chain | undefined = Chain | undefined,
+  TAccount extends SupportedAccounts = SupportedAccounts
+>({
+  account,
+  ...params
+}: UseSmartAccountClientProps<
+  TTransport,
+  TChain,
+  TAccount
+>): UseSmartAccountClientResult<TTransport, TChain, TAccount>;
+
+export function useSmartAccountClient({
+  account,
+  ...params
+}: UseSmartAccountClientProps): UseSmartAccountClientResult {
+  const bundlerClient = useBundlerClient();
+
+  switch (account.source) {
+    case "LightAccount":
+      return createAlchemySmartAccountClientFromRpcClient({
+        client: bundlerClient,
+        account,
+        ...params,
+      }).extend(lightAccountClientActions);
+    case "MultiOwnerModularAccount":
+      return createAlchemySmartAccountClientFromRpcClient({
+        client: bundlerClient,
+        account,
+        ...params,
+      })
+        .extend(multiOwnerPluginActions)
+        .extend(pluginManagerActions)
+        .extend(accountLoupeActions);
+    default:
+      throw new Error("Unsupported account type");
+  }
+}
```

Comment:
Before this impl, I wasn't using an overloaded function with the super general generics and I was having to do some typecasting of the `account` when creating the client so that typescript wouldn't yell at me. And even then, I was still getting type errors.

Adding this function overload where the generics are union types of the possible inputs / outputs allows you to define a typesafe api for the consumer while also maintaining type safety in the implementation without having to typecast. 

We do this in places like `createSmartAccountClient` too, but I don't think it really clicked until I needed to return to vastly different types like I do here. We similarly do that in https://github.com/alchemyplatform/aa-sdk/pull/514 which allows us to handle various EP versions here: https://github.com/alchemyplatform/aa-sdk/pull/544